src/link/MachO.zig

```diff
@@ -660,11 +652,18 @@ pub fn resolveLibSystem(
             "libSystem",
         )) break :success;
 
-        const dir = try comp.zig_lib_directory.join(tmp_arena, &[_][]const u8{ "libc", "darwin" });
-        const lib_name = try std.fmt.allocPrint(tmp_arena, "libSystem.{d}", .{
-            self.base.options.target.os.version_range.semver.min.major,
-        });
-        if (try accessLibPath(tmp_arena, &test_path, &checked_paths, dir, lib_name)) break :success;
+        switch (self.base.options.libc_provider) {
+            .none => unreachable,
+            .installation => unreachable,
+            .sysroot => {
+                const dir = try fs.path.join(tmp_arena, &[_][]const u8{ self.base.options.sysroot.?, "usr", "lib" });
+                if (try accessLibPath(tmp_arena, &test_path, &checked_paths, dir, "libSystem")) break :success;
+            },
+            .vendored => {
+                const dir = try comp.zig_lib_directory.join(tmp_arena, &[_][]const u8{ "libc", "darwin" });
+                if (try accessLibPath(tmp_arena, &test_path, &checked_paths, dir, "libSystem")) break :success;
+            },
+        }
```

Comment:
We have changed the logic here. Now, we allow `libSystem` in user search paths to take precedence over vendored or sysroot locations. Is that what we want?

Reply:
yes, this overall search scheme is desired; even for headers and frameworks:

1. user specified
2. (basically any other ways to specify that I don't list here)
3. (final) if vendored, search vendored, else search sysroot

so this change to `resolveLibSystem` is summarized as:

1.  pushes SDK/VENDORED to the end. Standard clang/gcc behavior is to put system stuff (ie. libs, headers) at end.
2. never mix SDK or VENDORED search dirs. I can't think of a case where it is desirable to _always_ search vendored. This solves 50% of #10957 wrt libs, but would need to check if headers are consistent too, for that issue).